### ISSUE TITLE
KOReader shows a strange splash image that leaves ghosting

### ISSUE THREAD

**Issue by zwim** [koreader/koreader#6923]
* KOReader version: koreader-android-arm-v2020.11-24-gc0b71d0_2020-11-29.apk 
(and on 2020-11-18;  2020-11-7. I have not tested others)
* Device: Tolino Epos2

#### Issue
KOReader shows the following splash image every time it is started (not only on installation). The image fills almost the whole screen.
After loading of a book is finished, there is unpleasant ghosting of that image on the screen:
![screencap](https://user-images.githubusercontent.com/36999612/100548533-6f982c80-326d-11eb-8dca-1eb8ffd2f8a9.png)

The ghosting does really _bother_ me. The image itself is not the problem.

#### Steps to reproduce
Start KOReader on Tolino Epos2. On my Android phone, there is the old small splash image (and of course no ghosting).

##### `crash.log` 
`adb logcat KOReader:I ActivityManager:* AndroidRuntime:* DEBUG:* *:F`.
[logcat.txt](https://github.com/koreader/koreader/files/5612508/logcat.txt)

##### Suggested solution 
If the new image is really wanted, then a full screen refresh should be done, before displaying a new book.

**Comment by pazos** [comment]
> The image fills almost the whole screen.

Caused by https://github.com/koreader/android-luajit-launcher/commit/067f8ab2588275c3f158134588ca4c4ae9e54839.

Either your device reports itself as a low dpi, which doesn't make any sense given its specs, or it is totally broken in a subtle way (it happened before). Anyhow I cannot reproduce it anywhere, it should be Tolino specific.

> KOReader shows the following splash image every time it is started (not only on installation). 

Yup, that happens from the start in https://github.com/koreader/android-luajit-launcher/pull/187. My only concern at that time was to avoid it onResume, so it only happens onCreate.

I'm totally fine to restrict its usage to assets extraction.

> After loading of a book is finished, there is unpleasant ghosting of that image on the screen

Since https://github.com/koreader/android-luajit-launcher/commit/623a834596c30a733b45352b2d704f01de513f59 we support `APP_CMD_WINDOW_REDRAW_NEEDED` callback. It was broken on frameworks before.

You can call [`framebuffer:_updateFull()`](https://github.com/koreader/koreader-base/blob/master/ffi/framebuffer_android.lua#L30) in https://github.com/koreader/koreader/blob/master/frontend/device/android/device.lua#L141 and see if that works (in that case leave `C.APP_CMD_GAINED_FOCUS` and `C.APP_CMD_INIT_WINDOW` with `this.device.screen:_updateWindow()` and use `this.device.screen:_updateFull()` only on `C.APP_CMD_WINDOW_REDRAW_NEEDED`, to prevent a lot of blinking)

tldr: except the last one all of the issues are specific to android-luajit-launcher. Better to check if is possible to circumvent device quirks without adding a different set of icons. About showing the splashscreen each time the program is launched: not needed, looks cool on lcd. I would be happy to remove it.

**Comment by Frenzie** [comment]
> About showing the splashscreen each time the program is launched: not needed, looks cool on lcd.

It does? It's a barely visible split second thing except on resource extraction after an update. :-P

**Comment by pazos** [comment]
> It does? It's a barely visible split second thing except on resource extraction after an update. :-P

I guess it depends on the animation used by the program that starts the activity. It is very visible on my phone.

**Comment by zwim** [comment]
@pazos: Without knowing what i am doing, it was more a try and see.

device.lua L139
I have commented out every updateWindow and updateFull, so far this sometimes works:
```
            elseif ev.code == C.APP_CMD_GAINED_FOCUS then
                this.device.screen:_updateWindow()
                this.device.screen:_updateFull()
            elseif ev.code == C.APP_CMD_INIT_WINDOW then
                this.device.screen:_updateWindow()
                this.device.screen:_updateFull()
            elseif ev.code == C.APP_CMD_WINDOW_REDRAW_NEEDED then
                this.device.screen:_updateWindow()
                this.device.screen:_updateFull()
```

Strange thing: I have installed the above and saw no ghosting. Then I closed and started KOReader again and the ghosting is there again.

**Comment by pazos** [comment]
@zwim: I'm afraid you'll need to build a debug version or simplify https://github.com/koreader/android-luajit-launcher/blob/master/jni/android_native_app_glue/android_native_app_glue.c#L34-L38 to https://github.com/koreader/android-luajit-launcher/blob/master/jni/android_native_app_glue/android_native_app_glue.c#L35.

Also, you'll need to add logger statements on each callback and figure out which comes first on your device. There're a million things happening on the shadows :/ 

If you had no problem with ghosting until recently that means that a proper icon size is enough to remove the ghosting you see with the huge image.

**Comment by pazos** [comment]
also, updateFull is updateWindow + eink, so there's no need to repeat them.

**Comment by pazos** [comment]
@zwim: does this help -> https://www.dropbox.com/s/aekp2xak8qf25ls/koreader-drawable.apk?dl=0 ??

**Comment by zwim** [comment]
@pazos: Strange. On installation I see a small (good looking, sharp) image. The splash says  96xx ms installation time. After it disapears the image remains.
On a restart only the image is shown, but nothing else.
![screencap](https://user-images.githubusercontent.com/36999612/100585706-466bb080-32ee-11eb-91bf-2451ade48d96.png)

logcat KOreader shows that it is working and opening some books, but the screen only shows that image :(

**Comment by pazos** [comment]
@zwim: very strange, indeed. You can build yourself an app with the changes in https://github.com/koreader/android-luajit-launcher/pull/264/commits/9e4631ca301aa302ddc5e2064e3d7b1685ed1bbd (just the file `app/res/drawable/splash_icon.xml`) and see if that alone is enough to get you a normal size icon without any other problem.

I'm afraid I cannot help here :/

**Comment by zwim** [comment]
@pazos: I have reverted the `this.device.screen:_updateFull()` statements on device.lua and applied the changes from 
`app/res/drawable/splash_icon.xml`.

Good new: I have absolutely no ghosting. The splash image is small and gets covered by the book opening info message.
Thanks a lot.
